File: src/pipeline.ts
import type { WebSocket } from "ws";
import { transcribe } from "./wyoming-stt.js";
import { synthesize } from "./wyoming-tts.js";
import { OpenClawClient } from "./openclaw-client.js";
import { ChannelServer } from "./channel-server.js";
import { nextSeq, getConfigFor } from "./device-registry.js";
import type { FollowUpMode } from "./device-config.js";
import { makeBinaryFrame } from "./utils.js";
import { getConfig } from "./config.js";
import { IdleSegmenter } from "./idle-segmenter.js";
import { SegmentQueue } from "./segment-queue.js";

const FOLLOW_UP_TAG = "[[follow_up]]";

function sendJSON(ws: WebSocket, obj: Record<string, unknown>): void {
  if (ws.readyState === ws.OPEN) {
    ws.send(JSON.stringify(obj));
  }
}

function sendAudioEnd(ws: WebSocket, followUp: boolean): void {
  sendJSON(ws, { type: "audio.end", follow_up: followUp });
}

function sendBinary(ws: WebSocket, deviceId: string, type: number, payload: Buffer): void {
  if (ws.readyState === ws.OPEN) {
    const seq = nextSeq(deviceId);
    ws.send(makeBinaryFrame(type, seq, payload));
  }
}

interface FollowUpResult {
  followUp: boolean;
  replyText: string;
}

export function resolveFollowUp(fullReply: string, mode: FollowUpMode): FollowUpResult {
  const trimmed = fullReply.trim();

  if (mode === "always") {
    return { followUp: true, replyText: stripFollowUpTag(trimmed) };
  }
  if (mode === "never") {
    return { followUp: false, replyText: stripFollowUpTag(trimmed) };
  }

  // "auto"
  if (trimmed.includes(FOLLOW_UP_TAG)) {
    return { followUp: true, replyText: stripFollowUpTag(trimmed) };
  }
  const trimmedEnd = trimmed.trimEnd();
  if (trimmedEnd.endsWith("?") || trimmedEnd.endsWith("？")) {
    return { followUp: true, replyText: trimmed };
  }
  return { followUp: false, replyText: trimmed };
}

function stripFollowUpTag(text: string): string {
  // Remove the tag plus any surrounding whitespace, then trim ends.
  return text.replace(/\s*\[\[follow_up\]\]\s*/g, " ").trim();
}

// In-stream variant: strip the follow_up tag but keep surrounding
// whitespace so consecutive flushed segments concatenate cleanly.
function stripFollowUpTagInline(text: string): string {
  return text.replace(/\s*\[\[follow_up\]\]\s*/g, " ");
}

async function synthesizeAndSend(ws: WebSocket, deviceId: string, text: string, signal: AbortSignal, targetRate?: number): Promise<void> {
  if (text.length > 0) {
    try {
      let sentStart = false;
      for await (const chunk of synthesize(text, {
        targetRate,
        signal,
        onSampleRate: (rate) => {
          if (!sentStart) {
            sendJSON(ws, { type: "audio.start", sample_rate: rate });
            sentStart = true;
          }
        },
      })) {
        if (signal.aborted) return;
        sendBinary(ws, deviceId, 0x02, chunk);
      }
    } catch (err) {
      console.error("[pipeline] TTS error:", (err as Error).message);
    }
  }
}

async function synthesizeError(ws: WebSocket, deviceId: string, signal: AbortSignal, targetRate?: number): Promise<void> {
  const errorMsg = "Sorry, I couldn't reach the backend. Please try again later.";
  try {
    let sentStart = false;
    for await (const chunk of synthesize(errorMsg, {
      targetRate,
      signal,
      onSampleRate: (rate) => {
        if (!sentStart) {
          sendJSON(ws, { type: "audio.start", sample_rate: rate });
          sentStart = true;
        }
      },
    })) {
      if (signal.aborted) return;
      sendBinary(ws, deviceId, 0x02, chunk);
    }
  } catch (ttsErr) {
    console.error("[pipeline] TTS error for error message:", (ttsErr as Error).message);
  }
}

function getFollowUpMode(deviceId: string): FollowUpMode {
  return getConfigFor(deviceId).follow_up_mode ?? "auto";
}

async function routeViaOpenClawDirect(
  deviceId: string,
  transcript: string,
  ws: WebSocket,
  openclawClient: OpenClawClient,
  signal: AbortSignal,
  targetRate?: number,
): Promise<void> {
  const sessionKey = `vauxr:${deviceId}`;
  let fullReply = "";

  try {
    await openclawClient.chat(sessionKey, transcript, (delta: string) => {
      fullReply = delta;
    });
  } catch (err) {
    sendJSON(ws, { type: "error", code: "BACKEND_ERROR", message: (err as Error).message });
    await synthesizeError(ws, deviceId, signal, targetRate);
    if (!signal.aborted) sendAudioEnd(ws, false);
    return;
  }

  if (signal.aborted) return;

  const { followUp, replyText } = resolveFollowUp(fullReply, getFollowUpMode(deviceId));
  console.log(`[pipeline] LLM reply (${replyText.length} chars, follow_up=${followUp}): ${replyText.substring(0, 200)}`);
  await synthesizeAndSend(ws, deviceId, replyText, signal, targetRate);
  if (!signal.aborted) sendAudioEnd(ws, followUp);
}

const CHANNEL_RESPONSE_TIMEOUT_MS = 60_000;

async function synthesizeSegment(
  ws: WebSocket,
  deviceId: string,
  text: string,
  signal: AbortSignal,
  targetRate: number | undefined,
  state: { sentStart: boolean },
): Promise<void> {
  if (text.length === 0) return;
  for await (const chunk of synthesize(text, {
    targetRate,
    signal,
    onSampleRate: (rate) => {
      if (!state.sentStart) {
        sendJSON(ws, { type: "audio.start", sample_rate: rate });
        state.sentStart = true;
      }
    },
  })) {
    if (signal.aborted) return;
    sendBinary(ws, deviceId, 0x02, chunk);
  }
}

async function routeViaChannel(
  deviceId: string,
  transcript: string,
  ws: WebSocket,
  channelServer: ChannelServer,
  signal: AbortSignal,
  targetRate?: number,
): Promise<void> {
  const sent = channelServer.sendTranscript(deviceId, transcript);
  if (!sent) {
    sendJSON(ws, { type: "error", code: "NO_CHANNEL", message: "Active channel not connected" });
    if (!signal.aborted) sendAudioEnd(ws, false);
    return;
  }
  console.log(`[pipeline] Awaiting channel response for ${deviceId}`);

  const idlePauseMs = getConfig().streamingTts.idlePauseMs;
  const startState = { sentStart: false };

  const queue = new SegmentQueue({
    synthesize: (text) => synthesizeSegment(ws, deviceId, text, signal, targetRate, startState),
    signal,
  });

  // Stream deltas through the idle-segmenter into the synth queue.
  // fullReply is accumulated independently for resolveFollowUp.
  let fullReply: string;
  try {
    fullReply = await new Promise<string>((resolve, reject) => {
      let accumulated = "";
      let resolved = false;

      const segmenter = new IdleSegmenter({
        idlePauseMs,
        onSegment: (segment) => {
          const cleaned = stripFollowUpTagInline(segment);
          if (cleaned.length > 0) queue.push(cleaned);
        },
        onEnd: (finalSegment) => {
          if (finalSegment !== null) {
            const cleaned = stripFollowUpTagInline(finalSegment);
            if (cleaned.length > 0) queue.push(cleaned);
          }
          queue.close();
        },
      });

      const timeout = setTimeout(() => {
        if (!resolved) {
          cleanup();
          segmenter.abort();
          queue.close();
          reject(new Error(`Channel response timeout after ${CHANNEL_RESPONSE_TIMEOUT_MS / 1000}s`));
        }
      }, CHANNEL_RESPONSE_TIMEOUT_MS);

      const cleanup = () => {
        resolved = true;
        clearTimeout(timeout);
        channelServer.removeResponseListener(deviceId);
      };

      channelServer.addResponseListener(deviceId, {
        onDelta: (_runId, text) => {
          if (resolved) return;
          accumulated += text;
          segmenter.push(text);
        },
        onEnd: (_runId) => {
          if (resolved) return;
          cleanup();
          segmenter.end();
          resolve(accumulated);
        },
        onError: (_runId, message) => {
          if (resolved) return;
          cleanup();
          segmenter.abort();
          queue.close();
          reject(new Error(message));
        },
      });

      signal.addEventListener("abort", () => {
        if (resolved) return;
        cleanup();
        segmenter.abort();
        queue.close();
        reject(new Error("Aborted"));
      }, { once: true });
    });
  } catch (err) {
    await queue.done();
    if (signal.aborted) return;
    sendJSON(ws, { type: "error", code: "BACKEND_ERROR", message: (err as Error).message });
    await synthesizeError(ws, deviceId, signal, targetRate);
    if (!signal.aborted) sendAudioEnd(ws, false);
    return;
  }

  // Wait for the synth worker to drain all flushed segments.
  await queue.done();
  if (signal.aborted) return;

  const { followUp, replyText } = resolveFollowUp(fullReply, getFollowUpMode(deviceId));
  console.log(`[pipeline] Channel reply (${replyText.length} chars, follow_up=${followUp}): ${replyText.substring(0, 200)}`);
  sendAudioEnd(ws, followUp);
}

export async function runVoiceTurn(
  deviceId: string,
  audioChunks: Buffer[],
  ws: WebSocket,
  openclawClient: OpenClawClient | null,
  channelServer: ChannelServer,
  signal: AbortSignal,
  targetRate?: number,
): Promise<void> {
  // Step 1: STT
  if (signal.aborted) return;

  let transcript: string;
  try {
    transcript = await transcribe(audioChunks);
  } catch (err) {
    console.error(`[pipeline] STT error for ${deviceId}:`, (err as Error).message);
    sendJSON(ws, { type: "error", code: "STT_ERROR", message: (err as Error).message });
    return;
  }

  if (signal.aborted) return;

  if (!transcript || transcript.trim().length === 0) {
    console.log(`[pipeline] Empty transcript for ${deviceId} — ending turn`);
    sendAudioEnd(ws, false);
    return;
  }

  console.log(`[pipeline] Transcript for ${deviceId}: "${transcript}"`);

  // Step 2: Send transcript to device
  sendJSON(ws, { type: "transcript", text: transcript });

  if (signal.aborted) return;

  // Step 3: Route to active channel
  const active = channelServer.getActiveChannel();

  if (active && active.type === "openclaw-direct" && openclawClient) {
    console.log(`[pipeline] Routing via openclaw-direct for ${deviceId}`);
    await routeViaOpenClawDirect(deviceId, transcript, ws, openclawClient, signal, targetRate);
  } else if (active && active.type !== "openclaw-direct") {
    console.log(`[pipeline] Routing via channel "${active.name}" (${active.type}) for ${deviceId}`);
    await routeViaChannel(deviceId, transcript, ws, channelServer, signal, targetRate);
  } else {
    console.warn("[pipeline] No active channel or backend available — dropping turn");
    sendJSON(ws, { type: "error", code: "NO_CHANNEL", message: "No active channel configured" });
    sendAudioEnd(ws, false);
  }
}

File: src/vauxr/pipeline.py
"""Voice-turn pipeline — STT → LLM (direct or channel) → TTS → device audio.

Port of `src/pipeline.ts`. The two routing modes (openclaw-direct vs.
channel plugin) keep the Node port's exact behavior: direct mode uses the
final cumulative delta as the full reply; channel mode accumulates
incremental deltas through the idle segmenter and TTS queue.
"""

from __future__ import annotations

import asyncio
import logging
import re
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from . import device_registry as registry
from .config import get_config
from .device_config import FollowUpMode
from .idle_segmenter import IdleSegmenter
from .protocol import encode_text_message
from .segment_queue import SegmentQueue
from .utils import make_binary_frame
from .wyoming_stt import transcribe
from .wyoming_tts import synthesize

if TYPE_CHECKING:
    from aiohttp.web import WebSocketResponse

    from .channel_server import ChannelServer
    from .openclaw_client import OpenClawClient

log = logging.getLogger("vauxr.pipeline")

_FOLLOW_UP_TAG = "[[follow_up]]"
_FOLLOW_UP_RE = re.compile(r"\s*\[\[follow_up\]\]\s*")
_CHANNEL_RESPONSE_TIMEOUT_S = 120.0
_ERROR_FALLBACK_TEXT = "Sorry, I couldn't reach the backend. Please try again later."


@dataclass(frozen=True)
class FollowUpResult:
    follow_up: bool
    reply_text: str


def strip_follow_up_tag(text: str) -> str:
    return _FOLLOW_UP_RE.sub(" ", text).strip()


def strip_follow_up_tag_inline(text: str) -> str:
    return _FOLLOW_UP_RE.sub(" ", text)


def resolve_follow_up(full_reply: str, mode: FollowUpMode) -> FollowUpResult:
    trimmed = full_reply.strip()
    if mode == "always":
        return FollowUpResult(follow_up=True, reply_text=strip_follow_up_tag(trimmed))
    if mode == "never":
        return FollowUpResult(follow_up=False, reply_text=strip_follow_up_tag(trimmed))
    # "auto"
    if _FOLLOW_UP_TAG in trimmed:
        return FollowUpResult(follow_up=True, reply_text=strip_follow_up_tag(trimmed))
    trimmed_end = trimmed.rstrip()
    if trimmed_end.endswith("?") or trimmed_end.endswith("？"):
        return FollowUpResult(follow_up=True, reply_text=trimmed)
    return FollowUpResult(follow_up=False, reply_text=trimmed)


def _follow_up_mode_for(device_id: str) -> FollowUpMode:
    mode = registry.get_config_for(device_id).get("follow_up_mode")
    return mode or "auto"


# --- WS helpers (the WS type is opaque; we duck-type on .send_str / .send_bytes / .closed) ---


async def _send_json(ws: Any, obj: dict[str, Any]) -> None:
    if getattr(ws, "closed", False):
        return
    try:
        await ws.send_str(encode_text_message(obj))
    except ConnectionResetError:
        pass


async def _send_binary(ws: Any, device_id: str, msg_type: int, payload: bytes) -> None:
    if getattr(ws, "closed", False):
        return
    seq = registry.next_seq(device_id)
    try:
        await ws.send_bytes(make_binary_frame(msg_type, seq, payload))
    except ConnectionResetError:
        pass


async def _send_audio_end(ws: Any, follow_up: bool) -> None:
    await _send_json(ws, {"type": "audio.end", "follow_up": follow_up})


async def _synthesize_and_send(
    ws: Any,
    device_id: str,
    text: str,
    abort: asyncio.Event,
    target_rate: int | None,
) -> None:
    if not text:
        return
    state = {"sent_start": False}

    def on_rate(rate: int) -> None:
        if not state["sent_start"]:
            asyncio.create_task(_send_json(ws, {"type": "audio.start", "sample_rate": rate}))
            state["sent_start"] = True

    try:
        async for chunk in synthesize(text, target_rate=target_rate, abort_event=abort, on_sample_rate=on_rate):
            if abort.is_set():
                return
            await _send_binary(ws, device_id, 0x02, chunk)
    except Exception as e:  # noqa: BLE001
        log.error("TTS error: %s", e)


async def _synthesize_error_message(
    ws: Any, device_id: str, abort: asyncio.Event, target_rate: int | None
) -> None:
    state = {"sent_start": False}

    def on_rate(rate: int) -> None:
        if not state["sent_start"]:
            asyncio.create_task(_send_json(ws, {"type": "audio.start", "sample_rate": rate}))
            state["sent_start"] = True

    try:
        async for chunk in synthesize(
            _ERROR_FALLBACK_TEXT, target_rate=target_rate, abort_event=abort, on_sample_rate=on_rate
        ):
            if abort.is_set():
                return
            await _send_binary(ws, device_id, 0x02, chunk)
    except Exception as e:  # noqa: BLE001
        log.error("TTS error for error message: %s", e)


# --- Routing ---


async def _route_via_openclaw_direct(
    device_id: str,
    transcript_text: str,
    ws: Any,
    openclaw_client: "OpenClawClient",
    abort: asyncio.Event,
    target_rate: int | None,
) -> None:
    session_key = f"vauxr:{device_id}"
    full_reply = ""

    def on_delta(text: str) -> None:
        nonlocal full_reply
        # openclaw-direct sends cumulative deltas — keep the latest as the
        # full reply (matches src/pipeline.ts routeViaOpenClawDirect).
        full_reply = text

    try:
        await openclaw_client.chat(session_key, transcript_text, on_delta)
    except Exception as err:  # noqa: BLE001
        await _send_json(
            ws, {"type": "error", "code": "BACKEND_ERROR", "message": str(err)}
        )
        await _synthesize_error_message(ws, device_id, abort, target_rate)
        if not abort.is_set():
            await _send_audio_end(ws, False)
        return

    if abort.is_set():
        return

    result = resolve_follow_up(full_reply, _follow_up_mode_for(device_id))
    log.info(
        "LLM reply (%d chars, follow_up=%s): %s",
        len(result.reply_text),
        result.follow_up,
        result.reply_text[:200],
    )
    await _synthesize_and_send(ws, device_id, result.reply_text, abort, target_rate)
    if not abort.is_set():
        await _send_audio_end(ws, result.follow_up)


async def _route_via_channel(
    device_id: str,
    transcript_text: str,
    ws: Any,
    channel_server: "ChannelServer",
    abort: asyncio.Event,
    target_rate: int | None,
) -> None:
    sent = channel_server.send_transcript(device_id, transcript_text)
    if not sent:
        await _send_json(
            ws, {"type": "error", "code": "NO_CHANNEL", "message": "Active channel not connected"}
        )
        if not abort.is_set():
            await _send_audio_end(ws, False)
        return

    log.info("Awaiting channel response for %s", device_id)
    idle_pause_ms = get_config().streaming_tts.idle_pause_ms
    start_state = {"sent_start": False}

    async def synth_segment(text: str) -> None:
        if not text:
            return

        def on_rate(rate: int) -> None:
            if not start_state["sent_start"]:
                asyncio.create_task(_send_json(ws, {"type": "audio.start", "sample_rate": rate}))
                start_state["sent_start"] = True

        async for chunk in synthesize(text, target_rate=target_rate, abort_event=abort, on_sample_rate=on_rate):
            if abort.is_set():
                return
            await _send_binary(ws, device_id, 0x02, chunk)

    queue = SegmentQueue(synthesize=synth_segment, abort_event=abort)

    # accumulated stores the full reply for resolve_follow_up.
    accumulated = ""
    response_done: asyncio.Future[str] = asyncio.get_event_loop().create_future()

    def on_segment_flush(text: str) -> None:
        cleaned = strip_follow_up_tag_inline(text)
        if cleaned:
            queue.push(cleaned)

    def on_segment_end(final_segment: str | None) -> None:
        if final_segment is not None:
            cleaned = strip_follow_up_tag_inline(final_segment)
            if cleaned:
                queue.push(cleaned)
        queue.close()

    segmenter = IdleSegmenter(
        idle_pause_ms=idle_pause_ms, on_segment=on_segment_flush, on_end=on_segment_end
    )

    # Channel listener — these are called from the channel server's reader.
    def on_delta(_run_id: str, text: str) -> None:
        nonlocal accumulated
        if response_done.done():
            return
        accumulated += text
        segmenter.push(text)

    def on_end(_run_id: str) -> None:
        if response_done.done():
            return
        segmenter.end()
        response_done.set_result(accumulated)

    def on_error(_run_id: str, message: str) -> None:
        if response_done.done():
            return
        segmenter.abort()
        queue.close()
        response_done.set_exception(RuntimeError(message))

    channel_server.add_response_listener(
        device_id, {"on_delta": on_delta, "on_end": on_end, "on_error": on_error}
    )

    # If abort fires, finish.
    abort_waiter = asyncio.create_task(abort.wait())

    def _on_abort(_t: asyncio.Task) -> None:
        if not response_done.done():
            segmenter.abort()
            queue.close()
            response_done.set_exception(RuntimeError("Aborted"))

    abort_waiter.add_done_callback(_on_abort)

    try:
        full_reply = await asyncio.wait_for(response_done, timeout=_CHANNEL_RESPONSE_TIMEOUT_S)
    except asyncio.TimeoutError:
        segmenter.abort()
        queue.close()
        await queue.done()
        abort_waiter.cancel()
        channel_server.remove_response_listener(device_id)
        if abort.is_set():
            return
        await _send_json(
            ws,
            {
                "type": "error",
                "code": "BACKEND_ERROR",
                "message": f"Channel response timeout after {int(_CHANNEL_RESPONSE_TIMEOUT_S)}s",
            },
        )
        await _synthesize_error_message(ws, device_id, abort, target_rate)
        if not abort.is_set():
            await _send_audio_end(ws, False)
        return
    except RuntimeError as err:
        await queue.done()
        abort_waiter.cancel()
        channel_server.remove_response_listener(device_id)
        if abort.is_set():
            return
        await _send_json(
            ws, {"type": "error", "code": "BACKEND_ERROR", "message": str(err)}
        )
        await _synthesize_error_message(ws, device_id, abort, target_rate)
        if not abort.is_set():
            await _send_audio_end(ws, False)
        return

    abort_waiter.cancel()
    channel_server.remove_response_listener(device_id)
    await queue.done()

    if abort.is_set():
        return

    result = resolve_follow_up(full_reply, _follow_up_mode_for(device_id))
    log.info(
        "Channel reply (%d chars, follow_up=%s): %s",
        len(result.reply_text),
        result.follow_up,
        result.reply_text[:200],
    )
    await _send_audio_end(ws, result.follow_up)


async def run_voice_turn(
    device_id: str,
    audio_chunks: list[bytes],
    ws: Any,
    openclaw_client: "OpenClawClient | None",
    channel_server: "ChannelServer",
    abort: asyncio.Event,
    target_rate: int | None = None,
) -> None:
    """Drive a full voice turn: STT, route to active channel, TTS, audio.end."""

    if abort.is_set():
        return

    try:
        transcript_text = await transcribe(audio_chunks)
    except Exception as err:  # noqa: BLE001
        log.error("STT error for %s: %s", device_id, err)
        await _send_json(ws, {"type": "error", "code": "STT_ERROR", "message": str(err)})
        return

    if abort.is_set():
        return

    if not transcript_text or not transcript_text.strip():
        log.info("Empty transcript for %s — ending turn", device_id)
        await _send_audio_end(ws, False)
        return

    log.info("Transcript for %s: %r", device_id, transcript_text)
    await _send_json(ws, {"type": "transcript", "text": transcript_text})

    if abort.is_set():
        return

    active = channel_server.get_active_channel()
    if active is not None and getattr(active, "type", None) == "openclaw-direct" and openclaw_client is not None:
        log.info("Routing via openclaw-direct for %s", device_id)
        await _route_via_openclaw_direct(
            device_id, transcript_text, ws, openclaw_client, abort, target_rate
        )
    elif active is not None and getattr(active, "type", None) != "openclaw-direct":
        log.info("Routing via channel %r for %s", getattr(active, "name", "?"), device_id)
        await _route_via_channel(device_id, transcript_text, ws, channel_server, abort, target_rate)
    else:
        log.warning("No active channel or backend available — dropping turn")
        await _send_json(
            ws, {"type": "error", "code": "NO_CHANNEL", "message": "No active channel configured"}
        )
        await _send_audio_end(ws, False)
